clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp

```diff
@@ -0,0 +1,181 @@
+	tags5_4,
+	tags5_5,
+};
+
+union union3 {
+	short *shorts;
+	int *ints;
+	float *floats;
+};
+
+union union4 {
+	short *shorts;
+	double *doubles;
+	int *ints;
+	float *floats;
+};
+
+// It is not obvious which enum is the tag for the union.
+struct taggedunion2 { // No warnings expected.
+	enum tags3 tagA;
+	enum tags4 tagB;
+	union union4 data;
+};
+
+// It is not obvious which union does the tag belong to.
+struct taggedunion4 { // No warnings expected.
+	enum tags3 tag;
+	union union3 dataB;
+	union union3 dataA;
+};
+
+struct taggedunion1 { // CHECK-MESSAGES: :[[@LINE]]:8: warning: Tagged union has more data members than tags! Data members: 4 Tags: 3 [bugprone-tagged-union-member-count]
+	enum tags3 tag;
+    union union4 data;
+};
+
+struct taggedunion5 { // CHECK-MESSAGES: :[[@LINE]]:8: warning: Tagged union has more data members than tags! Data members: 4 Tags: 3 [bugprone-tagged-union-member-count]
+	enum tags3 tag;
+    union {
+		int *ints;
+		char characters[13];
+		struct {
+			double re;
+			double im;
+		} complex;
+		long l;
+    } data;
+};
+
+struct taggedunion7 { // CHECK-MESSAGES: :[[@LINE]]:8: warning: Tagged union has more data members than tags! Data members: 4 Tags: 3 [bugprone-tagged-union-member-count]
+	enum {
+		tag1,
+		tag2,
+		tag3,
+	} tag;
+	union union4 data;
+};
+
+struct taggedunion8 { // CHECK-MESSAGES: :[[@LINE]]:8: warning: Tagged union has more data members than tags! Data members: 4 Tags: 3 [bugprone-tagged-union-member-count]
+	enum {
+		tag1,
+		tag2,
+		tag3,
+	} tag;
+	union {
+		int *ints;
+		char characters[13];
+		struct {
+			double re;
+			double im;
+		} complex;
+		long l;
+	} data;
+};
+
+struct nested1 { // CHECK-MESSAGES: :[[@LINE]]:8: warning: Tagged union has more data members than tags! Data members: 4 Tags: 3 [bugprone-tagged-union-member-count]
+	enum tags3 tag;
+	union {
+		char c;
+		short s;
+		int i;
+		struct { // CHECK-MESSAGES: :[[@LINE]]:3: warning: Tagged union has fewer data members than tags! Data members: 4 Tags: 5 [bugprone-tagged-union-member-count]
+			enum tags5 tag;
+			union union4 data;
+		} inner;
+	} data;
+};
+
+struct nested2 {
+	enum tags3 tag;
+	union {
+		float f;
+		int i;
+		struct { // CHECK-MESSAGES: :[[@LINE]]:3: warning: Tagged union has more data members than tags! Data members: 4 Tags: 3 [bugprone-tagged-union-member-count]
+			enum tags3 tag;
+			union union4 data;
+		} inner;
+	} data;
+};
+
+struct nested3 { // CHECK-MESSAGES: :[[@LINE]]:8: warning: Tagged union has fewer data members than tags! Data members: 2 Tags: 3 [bugprone-tagged-union-member-count]
+	enum tags3 tag;
+	union {
+		float f;
+		int i;
+		struct innerdecl { // CHECK-MESSAGES: :[[@LINE]]:10: warning: Tagged union has more data members than tags! Data members: 4 Tags: 3 [bugprone-tagged-union-member-count]
+			enum tags3 tag;
+			union union4 data;
+		}; 
+	} data;
+};
+
+enum tag_with_counter_lowercase {
+	node_type_loop,
+	node_type_branch,
+	node_type_function,
+	node_type_count,
```

Comment:
I might design my own language, where `count` can also be an AST node, then we report a false positive, right?

Reply:
Right, the heuristic can be disabled now in the configuration.